### PR TITLE
[DCP] Fix meta tensor loading

### PR DESCRIPTION
**Pull request by wz337**
We realized the fix for (https://github.com/pytorch/pytorch/pull/129683) loading the learning rate in place actually broke the meta tensor initialization. After the PR #129683, the learning rate is loading correctly, the param with meta tensors are still un-initialized.

We cannot use `tree_map_only_` to iterate over state_dict for initialization in-place,  as `empty_like` and `to("cuda")` are both not in-place option. More context in https://github.com/pytorch/pytorch/issues/130709 Therefore, with changes in (https://github.com/pytorch/pytorch/pull/129683), the tensor after loading are still meta tensors. We previously did not catch that since `self.assertEqual()` does not distinguish a DTensor with meta DTensor. 

In this PR, we added a _iterate_state_dict() function to implement in-place update for state_dict and updated the test to make sure that the params are no longer meta tensors after loading.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

**Comment by fegin** [comment]
LGTM, but can we let `state_dict_util._iterate_state_dict()` to support in place option so that we don't have to duplicate the code? Or we can add a TODO for the unification, lol.

**Comment by wz337** [comment]
> LGTM, but can we let `state_dict_util._iterate_state_dict()` to support in place option so that we don't have to duplicate the code? Or we can add a TODO for the unification, lol.

Sounds good. Added a TODO for the unification. We can follow up to make the changes but just want to fix the bug first.

**Review by fegin — APPROVED** [review]

**Review by fegin — APPROVED** [review]

**Review by fegin — APPROVED** [review]

File: torch/distributed/checkpoint/planner_helpers.py [hunk 1]
```diff
@@ -310,12 +307,80 @@ def _init_meta_tensor(value: Any) -> Any:
...
+        tensor_func,
+    )
+
+
+def _iterate_state_dict(
+    iter_object: Any,
+    dtensor_func: Callable,
+    sharded_tensor_func: Callable,
+    tensor_func: Callable,
+):
```
Comment (XuehaiPan) [inline torch/distributed/checkpoint/planner_helpers.py:347]: Since our use case is narrowed to nested `dict` rather than general pytrees, I think a better implementation is to have a `deep_update` function for nested `dict`s. Then you can reuse the `tree_map` function:

```python
nested_meta_tensor_dict = ...
initialized_nested_meta_tensor_dict = pytree.tree_map(init, nested_meta_tensor_dict)
deep_update(initialized_nested_meta_tensor_dict, nested_meta_tensor_dict)
```
